project-metrics/metrics_service/metric_plot.py

```diff
@@ -0,0 +1,80 @@
+import datetime
+import logging
+import matplotlib
+matplotlib.use('TkAgg')
+from matplotlib import pyplot as plt
+from typing import Sequence, Tuple
+import io
+
+from database import db
+from database import models
+from metrics import base as base_metric
+
+HISTORY_DAYS = 180
+
+
+class MetricHistoryPlotter(object):
+  """Plots metric results over the last six months.
+
+  WARNING: THIS CODE IS NOT THREAD-SAFE. DO NOT USE IN A REQUEST-HANDLER.
+  The underlying `matplotlib` library uses a shared TkAgg backend which only
+  works with one plot instance at a time. It is safe to call this from a Cron
+  task, but drawing in a web request context could result in parallel calls
+  drawing on each other.
+  """
+
+  def __init__(self, metric: base_metric.MetricImplementation):
+    self.metric = metric
+    self.history = self._get_metric_history()
+
+  def __del__(self):
+    plt.clf()
+
+  def plot_metric_history(self, width=4, height=2) -> io.BytesIO:
+    """Plots 6 months of metric results.
+
+    Args:
+      width: plot width in inches.
+      height: plot height in inches.
+
+    Returns:
+      A byte buffer containing the plot rendered as a PNG.
+    """
+    logging.info('Plotting %d days of results for %s', HISTORY_DAYS,
+                 self.metric.name)
+    self._make_plot()
+    self._set_labels()
+    return self._save_plot_to_buffer(width, height)
+
+  def _get_metric_history(self) -> Sequence[models.MetricResult]:
+    start_date = datetime.datetime.now() - datetime.timedelta(days=HISTORY_DAYS)
+
+    session = db.Session()
+    history = session.query(models.MetricResult).order_by(
+        models.MetricResult.computed_at.asc()).filter(
+            (models.MetricResult.name == self.metric.name)
+            & (models.MetricResult.computed_at > start_date)).all()
+
+    session.close()
+    return history
+
+  def _make_plot(self):
+    result_dates = [result.computed_at for result in self.history]
```

Comment:
Do these need to be lists? When dealing with buffers and lots of data it's worth trying to make these generators (`(…)`) instead of lists (`[…]`)

Reply:
Pretty sure matplotlib requires lists/converts generators to lists internally, as all docs talk in terms of sequences not iterables.